src/sas/system/config/config_meta.py

```diff
@@ -0,0 +1,183 @@
+
+        for key in data:
+
+            # Skip over any deleted attributes
+            if key in self._deleted_attributes:
+                continue
+
+            # Check the variable is in the schema
+            if key in self._schema:
+
+                try:
+                    coerced = self._schema[key].coerce(data[key])
+                    setattr(self, key, coerced)
+
+                except CoercionError as e:
+                    logger.error(f"Cannot set set variable '{key}', improper type ({e.message})")
+
+            else:
+                logger.error(f"Unknown config key: '{key}', skipping")
+
+    def save(self):
+        if self._write_disabled:
+            logger.error("Write disabled, this is probably because it will overwrite an outdated config.")
+            return
+
+        with open("config.json", 'w') as file:
+            self.save_to_file_object(file)
+
+    def save_to_file_object(self, file):
+        """ Save config file
+
+        Only changed variables will be included in the saved file
+        """
+        data = {}
+        for key in self._defaults:
+            old_value = self._defaults[key]
+            new_value = getattr(self, key)
+            if new_value != old_value:
+                data[key] = new_value
+
+        output_data = {
+            "sasview_version": sas.system.version.__version__,
+            "config_data": data}
+
+        json.dump(output_data, file)
+
+    def load(self):
+        filename = "config.json"
+        if os.path.exists(filename):
+            with open("config.json", 'r') as file:
+                self.load_from_file_object(file)
+
+    def load_from_file_object(self, file):
+        """ Load config file """
+        data = json.load(file)
+
+        if "sasview_version" not in data:
+            raise MalformedFile("Malformed config file - no 'sasview_version' key")
+
+        try:
+            parts = [int(s) for s in data["sasview_version"].split(".")]
+            if len(parts) != 3:
+                raise Exception
+
+        except Exception:
+            raise MalformedFile("Malformed version in config file, should be a string of the form 'X.Y.Z'")
+
+        if "config_data" not in data:
+            raise MalformedFile("Malformed config file - no 'config_data' key")
+
+        if not isinstance(data["config_data"], dict):
+            raise MalformedFile("Malformed config file - expected 'config_data' to be a dictionary")
+
+
+        # Check major version
+        file_version = data["sasview_version"]
+        file_major_version = file_version.split(".")[0]
+        sasview_major_version = sas.system.version.__version__.split(".")[0]
+
+        if int(file_major_version) != int(sasview_major_version):
+            logger.warning(f"Attempting to used outdated config file (config is"
+                           f" for {file_version}, this SasView version is {sas.system.version.__version__})")
+            self._write_disabled = True
```

Comment:
It **should** save the config in the startup path, so different versions wont share the same config.

> can you open an newer version of the config file with an older version of sasview

No. It is backward compatible, not forward compatible, as you say. It really shouldn't be forward compatible. The warning is just there for developers who will likely change the major version, but keep the location of the code the same.

The point is *exactly* that old keys shouldn't be hanging around. I spent two weeks finding all the old config variables that were no longer use and were just hanging around, messing up the code. The whole point of this PR is to minimise the accumulation of junk config variables, and make it easier to delete them.

Reply:
The use case for old and new versions of sasview coexisting in the machine is real. Scientists are distrustful sorts, and don't want to commit to the new without comparing to the old.

I was assuming that `config.json` was loaded/saved from somewhere in `~/.sasview`. Looking at the code in `config_meta.py` it appears to be loaded/saved from the current working directory?!?

I don't see any protection against version collision except as an accident of the O/S startup context. This seems error-prone. And I don't think we want protection from version collision. If I'm dutifully keeping sasview up to date I don't want to reset my configuration each time there is a new version.

Furthermore, I'm specifically suggesting that you do want forward and backward compatibility. You can do this by retaining all the keys in the config file for round-tripping, ignoring new keys in the old version and ignoring old keys in the new version. I'm not suggesting that you maintain access properties for these keys in the config structure. The remainder of the code will only see the available keys for that version. Missing keys (i.e., loading an old config in a new version) should be filled with default values, and saved to the config file on exit. These will be ignored when loading the config from the older version.

Reply:
> I was assuming that config.json was loaded/saved from somewhere in ~/.sasview. Looking at the code in config_meta.py it appears to be loaded/saved from the current working directory?!?

See my comment above

> Furthermore, I'm specifically suggesting that you do want forward and backward compatibility. You can do this by retaining all the keys in the config file for round-tripping, ignoring new keys in the old version and ignoring old keys in the new version. I'm not suggesting that you maintain access properties for these keys in the config structure. The remainder of the code will only see the available keys for that version. Missing keys (i.e., loading an old config in a new version) should be filled with default values, and saved to the config file on exit. These will be ignored when loading the config from the older version.

The problem with that, one which I was explicitly trying to avoid, is that it won't track changes to default values correctly. If you change a default value with the system you suggest, everyone has to update their config file to see the default behaviour. With the system as I have implemented it, if you have the default behaviour selected and that default changes, your version will reflect that. It's a choice with pros and cons, but my experience is that keeping every developers version as in-sync as possible makes things much easier in the long run.

Another the problem with keeping keys around is that it only takes someone to ignore a variable like `SHOW_THINGY` and make something very similar, `SHOW_THING` to do the same or a similar thing. Now there's two with almost the same name, nobody can remember which one was which, people complain that setting the variable does nothing because they've just changed the first one they see, and nobody wants to fix it because its loads of effort to find out what each one does. This is the motivation behind having some kind of versioning - it's a built in opportunity to fix this sort of thing. Whether or not it needs to complain so hard about the wrong version of the file is something that remains to be tested in practice, but as I mentioned before, it probably makes sense to have a different config file for each major version anyway.

In this PR, specifying an invalid key doesn't break anything, it just gives you log message. It will, however, forget it - I can see how that could potentially be a problem in the future. I will have it retain unknown keys and write them out afterwards.

Reply:
>  it won't track changes to default values correctly.

This is why I specified that "you should never change how they are used" 8-)

To support changing defaults your can capture the stored keys on load and tweak the save so that the value is stored if it was there before or if it is different from the default:
```python
def load_from_file_object(self, file):
    data = json.load(file)
    self._stored_keys = set(data)
    ...

def save_to_file_object(self):
    data = {
        key: value for key in self._defaults for value in [getattr(self, key)]
        if key in self._stored_keys or value != self._defaults[key]
    }
    ...
```

Reply:
That won't quite work right, but I will do something similar.

Reply:
Oops… this doesn't preserve unused keys. Try:
```python
def load_from_file_object(self, file):
    data = json.load(file)
    self._stored = data
    ...

def save_to_file_object(self):
    changed = {key: value for key, default in self._defaults.items() if default != (value := getattr(self, key))}
    data = {**self._stored, **changed}
    ...
```

This is different from the behaviour of your algorithm when using a single version. With your code if you turn on OpenCL then it will save the "use opencl" flags, but if you turn it off again it will match the default and save nothing. This new algorithm will preserve the "don't use opencl" flags when you turn it off. This is only the case if you turn it off in a different session than you turn it on. If it is the same session then the "on" values won't be in stored, so turning it off again won't trigger stored test or the non-default test.

A quick check shows that the key order in stored is preserved in the dict union constructor, though I'm not sure it matters since we are not round-tripping with preserved formatting and comments. And since we are not preserving formatting, we may want to sort keys as well as pretty-print on dump for debugging convenience.